cmake/StaticBuild.cmake

```diff
@@ -134,6 +164,10 @@ if(BUILD_STATIC_GCC)
   endif()
   if(WIN32)
     set(STATIC_BASE_LIBRARIES "${STATIC_BASE_LIBRARIES} -lmingw32 -lgcc_eh -lgcc -lmoldname -lmingwex -lmsvcrt")
+    if(CMAKE_GENERATOR STREQUAL "MinGW Makefiles")
+      # pthread has to be statically linked after libraries above and before kernel32
+      set(STATIC_BASE_LIBRARIES "${STATIC_BASE_LIBRARIES} -Wl,-Bstatic -lpthread -Wl,-Bdynamic")
```

Comment:
I still don't understand where this requirement comes from. pthread is not something that should be used on Windows. What error does this fix?

Reply:
Here are the first relevant errors from my build without lpthread:
```
[ 59%] Linking CXX executable vncconfig.exe
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.1.0/../../../../lib\libgnutls.a(system.o):(.text+0x6): undefined reference to `clock_gettime'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.1.0\libstdc++.a(eh_alloc.o):(.text$__tcf_0+0xc): undefined reference to `pthread_mutex_destroy'
...
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.1.0/libgcc_eh.a(emutls.o):(.text+0x5e): undefined reference to `pthread_mutex_init'
...
collect2.exe: error: ld returned 1 exit status
mingw32-make[2]: *** [win\vncconfig\CMakeFiles\vncconfig.dir\build.make:150: win/vncconfig/vncconfig.exe] Error 1
mingw32-make[1]: *** [CMakeFiles\Makefile2:637: win/vncconfig/CMakeFiles/vncconfig.dir/all] Error 2
mingw32-make: *** [Makefile:149: all] Error 2
```
It looks like MSYS2 requires pthread for libgnutls.a, libstdc++.a and libgcc_eh.a.  I have found many MSYS2 static libraries introduce many additional dependencies (including dynamic library dependencies).

Reply:
I found a cmake function to check for pthreads.  In the conversation view it was not updating for me.  However, if you check the actual changes the pthread library is only linked if cmake finds it is used.  The issue here is that msys2 mingw-w64 compiler compiles pthreads into gcc and the accompanying libraries.